V4L2 codec interface

6by9 opened the pull request:
Pushing for comment.

The first 11 patches were submitted by Stefan to the list (25th Oct - not in the archives!) but a couple of changes were requested. We could wait for those changes or take as is. Where I've based stuff on his I've taken on the comments made on his patches.
I need ```vchiq_register_child``` to be able to add the codec driver and vcsm-cma correctly with dma masks (I do get given the wrong physical address alias, so something is wrong that I haven't got to the bottom of).

The last patch is being resurrected on linux-media with an updated patch being pushed today (not yet archived). It makes life significantly easier for me as the buffer is exported within the driver to pass to vcsm-cma and be mem_wrapped, and otherwise a standard shutdown can trigger warnings as the VPU hasn't released the buffer in time. All buffers are ref counted via dmabufs, so it is safe.

The rest is all up for review.

Commit:
media: vb2: Allow reqbufs(0) with "in use" MMAP buffers

Videobuf2 presently does not allow VIDIOC_REQBUFS to destroy outstanding
buffers if the queue is of type V4L2_MEMORY_MMAP, and if the buffers are
considered "in use".  This is different behavior than for other memory
types and prevents us from deallocating buffers in following two cases:

1) There are outstanding mmap()ed views on the buffer. However even if
   we put the buffer in reqbufs(0), there will be remaining references,
   due to vma .open/close() adjusting vb2 buffer refcount appropriately.
   This means that the buffer will be in fact freed only when the last
   mmap()ed view is unmapped.

2) Buffer has been exported as a DMABUF. Refcount of the vb2 buffer
   is managed properly by VB2 DMABUF ops, i.e. incremented on DMABUF
   get and decremented on DMABUF release. This means that the buffer
   will be alive until all importers release it.

Considering both cases above, there does not seem to be any need to
prevent reqbufs(0) operation, because buffer lifetime is already
properly managed by both mmap() and DMABUF code paths. Let's remove it
and allow userspace freeing the queue (and potentially allocating a new
one) even though old buffers might be still in processing.

Signed-off-by: John Sheu <sheu@chromium.org>
Reviewed-by: Pawel Osciak <posciak@chromium.org>
Reviewed-by: Tomasz Figa <tfiga@chromium.org>
Signed-off-by: Tomasz Figa <tfiga@chromium.org>

6by9 committed:
staging: bcm2835-camera: Correctly denote key frames in encoded data

Forward MMAL key frame flags to the V4L2 buffers.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Return early on errors

Fix several instances where it is easier to return
early on error conditions than handle it as an else
clause.
As requested by Mauro.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Remove dead email addresses

None of the listed author email addresses were valid.
Keep list of authors and the companies they represented.
Update my email address.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Do not bulk receive from service thread

vchi_bulk_queue_receive will queue up to a default of 4
bulk receives on a connection before blocking.
If called from the VCHI service_callback thread, then
that thread is unable to service the VCHI_CALLBACK_BULK_RECEIVED
events that would enable the queue call to succeed.

Add a workqueue to schedule the call vchi_bulk_queue_receive
in an alternate context to avoid the lock up.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix comment style violations.

Fix comment style violations in the header files.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix spacing around operators

Fix checkpatch warnings over spaces around operators.
Many were around operations that can be replaced with the
BIT(x) macro, so replace with that where appropriate.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Reduce length of enum names

We have numerous lines over 80 chars, or oddly split. Many
of these are due to using long enum names such as
MMAL_COMPONENT_CAMERA.
Reduce the length of these enum names.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix multiple line dereference errors

Fix checkpatch errors "Avoid multiple line dereference"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix brace style issues.

Fix mismatched or missing brace issues flagged by checkpatch.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix missing lines between items

Fix checkpatch errors for missing blank lines after variable
or structure declarations.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix logical continuation splits

Fix checkpatch errors for "Logical continuations should be
on the previous line".

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix open parenthesis alignment

Fix checkpatch "Alignment should match open parenthesis"
errors.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Set sequence number correctly

Set the sequence number in vb2_v4l2_buffer mainly so the
latest v4l2-ctl reports the frame rate correctly.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835_camera: Ensure all buffers are returned on disable

With the recent change to match MMAL and V4L2 buffers there
is a need to wait for all MMAL buffers to be returned during
stop_streaming.

Fixes: 9384167 "staging: bcm2835-camera: Remove V4L2/MMAL buffer remapping"
Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Remove check of the number of buffers supplied

Before 9384167 there was a need to ensure that there were sufficient
buffers supplied from the user to cover those being sent to the VPU
(always 1).
With 9384167 the buffers are linked 1:1 between MMAL and V4L2,
therefore there is no need for that check, and indeed it is wrong
as there is no need to submit all the buffers before starting streaming.

Fixes: 9384167 "staging: bcm2835-camera: Remove V4L2/MMAL buffer remapping"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Handle empty EOS buffers whilst streaming

The change to mapping V4L2 to MMAL buffers 1:1 didn't handle
the condition we get with raw pixel buffers (eg YUV and RGB)
direct from the camera's stills port. That sends the pixel buffer
and then an empty buffer with the EOS flag set. The EOS buffer
wasn't handled and returned an error up the stack.

Handle the condition correctly by returning it to the component
if streaming, or returning with an error if stopping streaming.

Fixes: 9384167 "staging: bcm2835-camera: Remove V4L2/MMAL buffer remapping"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Ensure timestamps never go backwards.

There is an awkward situation with H264 header bytes. Currently
they are returned with a PTS of 0 because they aren't associated
with a timestamped frame to encode. These are handled by either
returning the timestamp of the last buffer to have been received,
or in the case of the first buffer the timestamp taken at
start_streaming.
This results in a race where the current frame may have started
before we take the start time, which results in the first encoded
frame having an earlier timestamp than the header bytes.

Ensure that we never return a negative delta to the user by checking
against the previous timestamp.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Add multiple inclusion protection to headers

mmal-common.h and mmal-msg.h didn't have the normal
ifndef FOO / define FOO / endif protection to stop it being
included multiple times. Add it.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc04_services: Split vchiq-mmal into a module

In preparation for adding a video codec V4L2 module which also
wants to use vchiq-mmal functions, split it out into an
independent module.
The minimum number of changes have been made to achieve this
(eg straight moves where possible) so existing checkpatch
errors will still be present.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: mmal-vchiq: Allocate and free components as required

The existing code assumed that there would only ever be 4 components,
and never freed the entries once used.
Allow arbitrary creation and destruction of components.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: mmal-vchiq: Make timeout a defined parameter

The timeout period for VPU communications is a useful thing
to extend when debugging.
Set it via a define, rather than a magic number buried in the code.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: mmal-vchiq: Make a mmal_buf struct for passing parameters

The callback from vchi_mmal to the client was growing lots of extra
parameters. Consolidate them into a single struct instead of
growing the list further.
The struct is associated with the client buffer, therefore there
are various changes to setup various containers for the struct,
and pass the appropriate members.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: mmal-vchiq: Add support for event callbacks.

(Preparation for the codec driver).
The codec uses the event mechanism to report things such as
resolution changes. It is signalled by the cmd field of the buffer
being non-zero.

Add support for passing this information out to the client.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc04_services: Support sending data to MMAL ports

Add the ability to send data to ports. This only supports
zero copy mode as the required bulk transfer setup calls
are not done.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc04_services: Add new vc-sm-cma driver

This new driver allows contiguous memory blocks to be imported
into the VideoCore VPU memory map, and manages the lifetime of
those objects, only releasing the source dmabuf once the VPU has
confirmed it has finished with it.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc04_services: Add a V4L2 M2M codec driver

This adds a V4L2 memory to memory device that wraps the MMAL
video decode and video_encode components for H264 and MJPEG encode
and decode, MPEG4, H263, and VP8 decode (and MPEG2 decode
if the appropriate licence has been purchased).

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc04_services: Use vc-sm-cma to support zero copy

With the vc-sm-cma driver we can support zero copy of buffers between
the kernel and VPU. Add this support to vchiq-mmal.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc04_services: Fixup vchiq-mmal include ordering

There were dependencies on including the headers in the correct
order. Fix up the headers so that they include the other
headers that they depend on themselves.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

nathanchance committed:
staging: bcm2835-camera: Avoid unneeded internal declaration warning

Clang warns:

drivers/staging/vc04_services/bcm2835-camera/controls.c:59:18: warning:
variable 'mains_freq_qmenu' is not needed and will not be emitted
[-Wunneeded-internal-declaration]
static const s64 mains_freq_qmenu[] = {
                 ^
1 warning generated.

This is because mains_freq_qmenu is currently only used in an ARRAY_SIZE
macro, which is a compile time evaluation in this case. Avoid this by
adding mains_freq_qmenu as the imenu member of this structure, which
matches all other controls that uses the ARRAY_SIZE macro in v4l2_ctrls.
This turns out to be a no-op because V4L2_CID_MPEG_VIDEO_BITRATE_MODE is
defined as a MMAL_CONTROL_TYPE_STD_MENU, which does not pass the imenu
definition along to v4l2_ctrl_new in bm2835_mmal_init_controls.

Link: https://github.com/ClangBuiltLinux/linux/issues/122
Signed-off-by: Nathan Chancellor <natechancellor@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

6by9 committed:
staging: bcm2835-camera: Use enums for max value in controls

Controls of type MMAL_CONTROL_TYPE_STD_MENU call v4l2_ctrl_new_std_menu
with a max value and a mask. The max value is one of the defined
values for the control, however in the config array there are several
entries where raw numbers have been used instead. Replace these
with the appropriate enum.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Remove/amend some obsolete comments

Remove a todo which has been done.
Remove a template line that was redundant.
Make a comment clearer as to the non-obvious meaning of a field.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Correct V4L2_CID_COLORFX_CBCR behaviour

With V4L2_CID_COLORFX_CBCR calling ctrl_set_colfx it was incorrectly
assigning the colour values to the enable field of dev->colourfx
instead of the u and v fields.

Correct the assignments.

Reported as a Coverity issue
Detected by CoverityScan CID#1419711 ("Unused value")

Reported-by: Colin Ian King <colin.king@canonical.com>
Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

Commit:
staging: bcm2835-audio: Enable compile test

Enable the compilation test for bcm2835-audio.

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: vchiq_arm: Fix platform device unregistration

In error case platform_device_register_data would return an ERR_PTR
instead of NULL. So we better check this before unregistration.

Fixes: 37b7b3087a2f ("staging/vc04_services: Register a platform device for the camera driver.")
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835-audio: use module_platform_driver() macro

There is not much value behind this boilerplate, so use
module_platform_driver() instead.

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835-audio: Drop DT dependency

Just like the bcm2835-video make this a platform driver which is probed
by vchiq. In order to change the number of channels use a module
parameter instead, but use the maximum as default.

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835-camera: fix module autoloading

In order to make the module bcm2835-camera load automatically, we need to
add a module alias.

Fixes: 4bebb0312ea9 ("staging/bcm2835-camera: Set ourselves up as a platform driver.")
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835-camera: Move module info to the end

In order to have this more consistent between the vc04 services move
the module information to the end of the file.

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: vchiq_arm: Register a platform device for the audio driver

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835-camera: Provide more specific probe error messages

Currently there is only a catch-all info message which print the
relevant error code without any context. So add more specific error
messages in order to narrow down possible issues.

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835-camera: Add hint about possible faulty GPU mem config

As per default the GPU memory config of the Raspberry Pi isn't sufficient
for the camera usage. Even worse the bcm2835 camera doesn't provide a
helpful error message in this case. So let's add a hint to point the user
to the likely cause.

Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: bcm2835: Don't probe if no camera is detected

It is a waste of resources to load the camera driver in case there isn't
a camera actually connected to the Raspberry Pi. This solution also
avoids a NULL ptr dereference of mmal instance on driver unload.

Fixes: 7b3ad5abf027 ("staging: Import the BCM2835 MMAL-based V4L2 camera driver.")
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

Commit:
staging: vchiq_arm: Fix camera device registration

Since the camera driver isn't probed via DT, we need to properly setup DMA.

Fixes: 37b7b3087a2f ("staging/vc04_services: Register a platform device for the camera driver.")
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>

6by9 committed:
staging: vchiq_arm: Register bcm2835-codec as a platform driver

Following the same pattern as bcm2835-camera and bcm2835-audio,
register the V4L2 codec driver as a platform driver

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vchiq_arm: Register vcsm-cma as a platform driver

Following the same pattern as bcm2835-camera and bcm2835-audio,
register the vcsm-cma driver as a platform driver

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Ensure H264 header bytes get a sensible timestamp

H264 header come from VC with 0 timestamps, which means they get a
strange timestamp when processed with VC/kernel start times,
particularly if used with the inline header option.
Remember the last frame timestamp and use that if set, or otherwise
use the kernel start time.

https://github.com/raspberrypi/linux/issues/1836

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
ARM: bcm2835_defconfig: Enable bcm2835-camera

Enables the V4L2 camera driver as a module.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix alignment should match open parenthesis

Fix up checkpatch "Alignment should match open parenthesis" errors

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix multiple assignments should be avoided

Clear checkpatch complaints of "multiple assignments should be avoided"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Fix up all formatting in mmal-paramters.h

Fixes up all checkpatch errors in mmal-parameters.h

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: mmal-vchiq: Avoid use of bool in structures

Fixes up a checkpatch error "Avoid using bool structure members
because of possible alignment issues".

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
ARM: bcm2835_defconfig: Enable bcm2835-codec

Enables the V4L2 M2M codec driver as a module.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
media: videobuf2: Allow exporting of a struct dmabuf

videobuf2 only allowed exporting a dmabuf as a file descriptor,
but there are instances where having the struct dma_buf is
useful within the kernel.

Split the current implementation into two, one step which
exports a struct dma_buf, and the second which converts that
into an fd.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: vc-sm-cma: Fixup driver for older VCHI APIs

Original patch was based off staging which included some cleanups
of the VCHI APIs. Those aren't present here, so switch back to
the older API.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
config: Add bcm2835-codec to Pi defconfigs.

Adds the V4L2 M2M codec driver to the config.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

popcornmix commented:
> We could wait for those changes or take as is.

As long as upstream versions have (approximately) the same commit message and functionality, then no problem with merging these now. When commits appear upstream we can just drop the downstream commits on the rebase.

6by9 commented:
Decoder works quite happily via GStreamer 1.14. Not tested on 1.10(?) which is shipped with Stretch.
```
gst-launch-1.0 -e -vvv filesrc location=usbhdd/big_buck_bunny_1080p_h264.mov ! qtdemux name=demux ! queue ! h264parse ! v4l2h264dec capture-io-mode=4 ! kmssink demux. ! queue ! faad ! autoaudiosink
```
plays back Big Buck Bunny via the full KMS driver pretty well (audio and video, but a few jerks in the video).

Transcoding is working
```
gst-launch-1.0 -e -vvv filesrc location=usbhdd/big_buck_bunny_1080p_h264.mov ! qtdemux name=demux ! queue ! h264parse ! v4l2h264dec ! v4l2h264enc ! filesink location=foo.h264
```
You should be able to alter the encoding parameters too although I've not tested it.

Plain encoding from v4l2src is giving some errors on UYVY - need to investigate further.

6by9 commented:
> As long as upstream versions have (approximately) the same commit message and functionality, then no problem with merging these now. When commits appear upstream we can just drop the downstream commits on the rebase.

The request was a simple one to check for an error from ```vchiq_register_child``` during probe and set the handle to NULL on error, rather than having to check ```IS_ERR``` during ```vchiq_remove```. That's why mine do that whilst Stefan's don't.
I'd expect the commit text to remain the same.

popcornmix commented:
I think stretch gstreamer is too old (it failed for me), but I could play video with ffmpeg-drm test app.
The kernel didn't obviously break anything (kodi with the legacy kms driver still plays hevc for example).
I didn't spot anything on a quick scan through the commits so no objections here.

6by9 commented:
I'm not in tomorrow morning, but will have a look at what level of working I can achieve with GStreamer in Stretch. Buster has GStreamer 1.14.4, so that should support it.

Next job is to get Chromium built and trying to use it in the GL world. Not quite so trivial as they use custom kernel headers which aren't then shipped in the Chromium source :-/ It also uses some hardcoded node names - /dev/video-dec and /dev/video-enc.
I'll also be aiming to get these patches upstreamed - fun, fun, fun!

6by9 commented:
Rebased to 4.19.2.
Github appears to be doing odd things with the order it is listing the commits, but the hashes are correct.

6by9 commented:
Rebased again to 4.19.4 with minor fix to the encode path for YUYV or RGB formats.

Tested with GStreamer 1.15 again. GST1.10 appears to be just too old to be of any real use.
Also tested with FFmpeg and seems to do sensible things on at least decoding.

I'm trying to get Chromium running with Buster as it has the latest libDRM and Mesa.

What else do you want to be done before this is merged?

6by9 committed:
staging: bcm2835-camera: Fix stride on RGB3/BGR3 formats

RGB3/BGR3 end up being 3 bytes per pixel, which meant that
the alignment code ended up trying to align using bitmasking
with a mask of 96.
That doesn't work, so switch to an arithmetic alignment for
those formats.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
@pelwell Any chance of a once over?

Chromium is proving to be a pain as it pulls an alternate huge lump of UI code, which seems to not work at the moment (segfaults on starting). The current build doesn't use any of that, so John hasn't any pearls of wisdom to impart. I'll keep poking at it, but it'd be nice to get this merged (rebasing the PR with each kernel bump gets boring).

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -358,68 +360,72 @@ static void buffer_cb(struct vchiq_mmal_instance *instance,
+
+	vb2_set_plane_payload(&buf->vb.vb2_buf, 0, length);
+	if (mmal_flags & MMAL_BUFFER_HEADER_FLAG_KEYFRAME)
+		buf->vb.flags |= V4L2_BUF_FLAG_KEYFRAME;
```
Where did this pair of lines appear from?

6by9 replied:
Should have been a cherry-pick of 5b0d431bd87468f13f7ba8c96eb791c4c4bee467 "BCM2835-V4L2: Correctly denote key frames in encoded data" from 4.14.
I'll find out what happened there.

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -799,27 +805,27 @@ static int vidioc_overlay(struct file *file, void *f, unsigned int on)
 
 	if (enable_camera(dev) < 0)
-		goto error;
+		return ret;
```
This doesn't actually set the return value, so the function will appear to succeed.

6by9 replied:
Ack.

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/mmal-common.h:
```diff
@@ -13,6 +13,8 @@
  *
  */
+#ifndef __MMAL_COMMON_H
+#define __MMAL_COMMON_H
```
Do you need the extra underscores here and not in mmal_msg.h?

6by9 replied:
We've a slight mismash with mmal_parameters.h having underscores with most of the others not. I'll unify them.

pelwell commented on drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.h:
```diff
@@ -82,6 +82,7 @@ struct vchiq_mmal_port {
 };
 
 struct vchiq_mmal_component {
+	u32 in_use:1;
```
Is this preferable to a bool (as featured on the line below)?

6by9 replied:
IIRC checkpatch whinges at bool. Looking at my ToT a later patch changes the line below to ```u32 enabled:1```

pelwell commented on drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.c:
```diff
@@ -44,6 +44,12 @@ MODULE_VERSION("0.0.1");
 #define VCHIQ_MMAL_MAX_COMPONENTS 64
 
+/*
+ * timeout for synchronous msg responses in seconds.
```
Capitalise `Timeout`, since it starts a sentence? I only noticed because of the line below.

6by9 replied:
Ack

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -267,14 +273,17 @@ static int buffer_init(struct vb2_buffer *vb)
+	buf->mmal.buffer = vb2_plane_vaddr(&buf->vb.vb2_buf, 0);
+	buf->mmal.buffer_size = vb2_plane_size(&buf->vb.vb2_buf, 0);
+
+	mmal_vchi_buffer_init(dev->instance, &buf->mmal);
```
Is this init function guaranteed never to fail? You don't mention this change,

6by9 replied:
Github loses the commit the reviewer was looking at when they made the comment, so it's possible this has become misplaced in the series.
A later patch does check the return value, but this one ought to too.

The vchi context is now attached to the buffer at allocation, not allocated on demand. I'll review to check that bits got added at the correct point.

6by9 replied:
(That makes more sense now that I've found your comment against a commit rather than in the summary. Blasted Github!)
Yes, the change was incorrect. I suspect it got tripped up in the port from 4.14 to 4.19 (buffer_init is already upstream in 4.19).

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -672,6 +690,14 @@ static void stop_streaming(struct vb2_queue *vq)
+		buf = container_of(vb2, struct vb2_mmal_buffer, vb);
+
+		mmal_vchi_buffer_cleanup(&buf->mmal);
+	}
```
Why has this loop suddenly become necessary?

6by9 replied:
As above, the vchi context is now attached to the buffer at allocation, not allocated on demand. I'll review to check that bits got added at the correct point.

6by9 replied:
You're right - this is handled from buffer_cleanup. I'm not sure why it got duplicated here.

pelwell reviewed (commented):
Batch 1 - up to "staging: mmal-vchiq: Make a mmal_buf struct for passing parameters"

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/controls.c:
```diff
@@ -1109,7 +1109,7 @@ static const struct bm2835_mmal_v4l2_ctrl v4l2_ctrls[V4L2_CTRL_COUNT] = {
 		V4L2_CID_POWER_LINE_FREQUENCY, MMAL_CONTROL_TYPE_STD_MENU,
 		0, ARRAY_SIZE(mains_freq_qmenu) - 1,
-		1, 1, NULL,
+		1, 1, mains_freq_qmenu,
```
It seems odd to include this patch when the next patch deletes mains_freq_qmenu.

6by9 replied:
The 1st patch got merged upstream, so I brought us into line with upstream before fixing the actual commit.

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -360,8 +360,13 @@ static void buffer_cb(struct vchiq_mmal_instance *instance,
 		}
```
Should you change the commit message from "H264 header come off VC" to "H264 headers come from VC"?

6by9 replied:
Ack

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -1062,12 +1062,15 @@ static int mmal_setup_components(struct bm2835_mmal_dev *dev,
 			    &dev->component[COMP_CAMERA]->output[CAM_PORT_VIDEO];
-		else
-			camera_port = port =
+			port = camera_port;
```
Move the duplicated assignment to port out outside the conditional and avoid the extra braces.

pelwell commented on drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c:
```diff
@@ -3656,6 +3657,9 @@ static int vchiq_probe(struct platform_device *pdev)
 	bcm2835_camera = vchiq_register_child(pdev, "bcm2835-camera");
 	bcm2835_audio = vchiq_register_child(pdev, "bcm2835_audio");
+	bcm2835_codec = vchiq_register_child(pdev, "bcm2835-codec");
+	if (IS_ERR(bcm2835_codec))
```
Why the inconsistent error handling? If you change the others (some of which were added in this patch set) to match in a later commit, is it not better to keep them consistent here then change them all?

6by9 replied:
Stefan's patchset has not been merged upstream as they wanted changes. Shall I make a fixup to his patch first include the error handling, and then add the new ones to match?

pelwell replied:
As you see fit, but that sounds preferable.

6by9 replied:
OK, will do.

pelwell commented on drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c:
```diff
@@ -3655,6 +3656,9 @@ static int vchiq_probe(struct platform_device *pdev)
 
+	vcsm_cma = vchiq_register_child(pdev, "vcsm-cma");
+	if (IS_ERR(vcsm_cma))
+		vcsm_cma = NULL;
```
Inconsistent error handling again.

6by9 replied:
As above, Stefan's patches aren't merged. I'll fix them up as a separate patch.

pelwell reviewed (commented):
2nd batch - the rest of the commits. Just a few minor quibbles.

pelwell commented:
GitHub is demanding another manual rebase, which I don't understand since there were no conflicts when I rebased the PR manually offline.

6by9 commented:
> GitHub is demanding another manual rebase, which I don't understand since there were no conflicts when I rebased the PR manually offline.

Each force push bumping the Linux release is wanting the manual rebase even with no conflicts. It's getting a little annoying, but not something we can fix.

pelwell commented:
I'll merge when you're happy with it.

6by9 commented:
Thanks for all your efforts in reviewing. I'll update as noted.

6by9 committed:
staging: vchiq_arm: Improve error handling on loading drivers

The handling of loading platform drivers requires checking IS_ERR
for the pointer on unload.
If the driver fails to load, NULL the pointer during probe as
platform_device_unregister already checks for NULL.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
staging: bcm2835-camera: Unify header inclusion defines

Most of the headers use ifndef FOO_H, whilst mmal-parameters.h
used ifndef __FOO_H.

Revise mmal-parameters.h to drop the underscores and make the
headers all consistent.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 reviewed (commented):
Hopefully the latest push should address all the issues you raised. Thanks again for the review.

pelwell commented:
I'm going to take your word for it and merge before another rebase is required
